VideoPress: render VideoChapters block on the backend

trakos opened the pull request:
#### Changes proposed in this Pull Request:
This PR adds rendering chapters block on the backend. The data is fetched from public-api, endpoint added in D91445-code

Currently, it uses a hard-coded GUID, so I'm marking it as a draft.

<img width="439" alt="image" src="https://user-images.githubusercontent.com/6437642/199633255-6644f8d1-fd4b-4b3e-9a2d-582a776c9ab4.png">

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
pe4Cmx-qW-p2

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
* Have D91445-code on your sandbox and set `JETPACK__SANDBOX_DOMAIN`.
* Follow instructions from https://github.com/Automattic/jetpack/pull/27241 to create a post with the VideoChapters block.
* Check that it is rendered when viewing the post - for now in a pretty simple style.

trakos committed:
VideoPress: render VideoChapters block on the backend

This PR adds rendering chapters block on the backend.
The data is fetched from public-api.

lhkowalski commented on projects/packages/videopress/src/class-chapters.php:
```diff
@@ -0,0 +1,66 @@
+		}
+
+		$html = '<ul>';
+		foreach ( $chapters['chapters']['en'] as $chapter ) {
```
In the future, the language (in this case, `en`) can be a parameter as well, or at least be determined by the site language.

lhkowalski replied:
`@todo` comment about it got added. I will resolve this for now.

trakos commented:
I'm closing this, since it was incorporated into https://github.com/Automattic/jetpack/pull/27253

retrofox commented:
> I'm closing this, since it was incorporated into #27253

we may re-open this PR. #27253 was created for development/testing/demo purposes. This PR should be ready to go to trunk.

retrofox commented on projects/packages/videopress/src/class-chapters.php:
```diff
@@ -0,0 +1,66 @@
+			array()
+		);
+
+		if ( is_wp_error( $result ) || 200 !== wp_remote_retrieve_response_code( $result ) ) {
+			wp_send_json_error( array( 'message' => __( 'Could not fetch VideoPress chapters. Please try again later.', 'jetpack-videopress-pkg' ) ) );
+
+			return array();
+		}
+
+		return json_decode( wp_remote_retrieve_body( $result ), true );
```
We need to handle the error in another way. Currently, it breaks the app, showing the error generated by the `wp_send_json_error` function. Maybe something like the following (another commit belong is needed too) :

```suggestion
		if ( is_wp_error( $result ) || 200 !== wp_remote_retrieve_response_code( $result ) ) {
			return array(
				'chapters' => array(),
				'error'    => true,
			);
		}

		return array(
			'chapters' => json_decode( wp_remote_retrieve_body( $result ), true ),
			'error'    => false,
		);
```

lhkowalski replied:
The suggestion is, if we fail while fetching the chapters, return an empty list of chapters and a flag signaling the failure. Is that correct?

retrofox replied:
correct. It's a suggestion, though

retrofox replied:
also you can send an error message instead of the `true` boolean value

retrofox commented on projects/packages/videopress/src/class-chapters.php:
```diff
@@ -0,0 +1,66 @@
+	 * @return string HTML markup
+	 */
+	public static function render_chapters( $video_guid ) {
+		$chapters = self::fetch_chapters( $video_guid );
```
Need to commit in case https://github.com/Automattic/jetpack/pull/27249/files#r1014478352 commits too
```suggestion
		$chaptersData = self::fetch_chapters( $video_guid );

		if ( $chaptersData['error'] ) {
			return '';
		}

		$chapters = $chaptersData['chapters'];
```

lhkowalski replied:
Added as well. Thanks!

lhkowalski committed:
Change the rendering to expect milliseconds on the start property

lhkowalski committed:
Change the return object on errors

In this new format, the client will receive an empty list of chapters (so we don't break anything) as well as a flag to inform that an error happened.

lhkowalski committed:
Handle request errors on the renderer method

lhkowalski committed:
Get the real video guid from block attributes

trakos committed:
Add `isset` check when checking for error

trakos commented:
Thing to note: it was suggested during review of related diff that this PR should be tested on wpcom simple once it's ready, before merging. We have to make sure that the API works there as well.
